Add file and line info to backtraces under Linux

yazd opened the pull request:
This work provides file and line info by reading the DWARF debug section `.debug_line` and resolving the stacktrace addresses.

Example stacktrace output on exception when program is compiled with `-g`:

```d
std.format.FormatException@/home/administrator/Projects/d-devel/phobos/std/format.d(487): Orphan format specifier: %s
----------------
/home/administrator/Projects/d-devel/phobos/std/exception.d:589 pure @safe bool std.exception.enforceEx!(std.format.FormatException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) [0x4335b7]
/home/administrator/Projects/d-devel/phobos/std/format.d:487 pure @safe uint std.format.formattedWrite!(std.array.Appender!(immutable(char)[]).Appender, char).formattedWrite(std.array.Appender!(immutable(char)[]).Appender, const(char[])) [0x430a09]
/home/administrator/Projects/d-devel/phobos/std/format.d:6548 pure @safe immutable(char)[] std.format.format!(char).format(const(char[])) [0x430473]
foo/bar.d:9 void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int) [0x430412]
foo/bar.d:19 void foo.bar.f2!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f2(int) [0x438c3c]
foo/bar.d:13 void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int) [0x430424]
foo/bar.d:19 void foo.bar.f2!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f2(int) [0x438c3c]
foo/bar.d:13 void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int) [0x430424]
main.d:10 _Dmain [0x4303e5]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x43f12e]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x43f089]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x43f0ea]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x43f089]
??:? _d_run_main [0x43f00a]
??:? main [0x43ad87]
??:? __libc_start_main [0x891eca3f]
```
vs
```d
std.format.FormatException@/usr/include/dmd/phobos/std/format.d(487): Orphan format specifier: %s
----------------
./main(pure @safe uint std.format.formattedWrite!(std.array.Appender!(immutable(char)[]).Appender, char).formattedWrite(std.array.Appender!(immutable(char)[]).Appender, const(char[]))+0x9a) [0x42c302]
./main(pure @safe immutable(char)[] std.format.format!(char).format(const(char[]))+0x34) [0x42bd94]
./main(void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int)+0x23) [0x42bd33]
./main(void foo.bar.f2!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f2(int)+0x15) [0x42f845]
./main(void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int)+0x35) [0x42bd45]
./main(void foo.bar.f2!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f2(int)+0x15) [0x42f845]
./main(void foo.bar.f1!(pure nothrow @nogc @safe int main.main().__lambda1(int)).f1(int)+0x35) [0x42bd45]
./main(_Dmain+0xe) [0x42bd06]
./main(_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv+0x1f) [0x435d47]
./main(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+0x2a) [0x435ca2]
./main(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll()+0x2b) [0x435d03]
./main(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+0x2a) [0x435ca2]
./main(_d_run_main+0x1d2) [0x435c22]
./main(main+0x20) [0x431a48]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f4f61981a40]
```
`-L--export-dynamic` linker option is still needed to provide the function names. Otherwise only file and line number are shown.

TODOs:
- [ ] add tests
- [x] add nothrow and @nogc
- [ ] ???

@complexmath, I added your name in authors field as I copied some code from rt/runtime.d.
@MartinNowak, you were interested in this.

Please review freely and if you see any possible enhancements, I will work on them.

Edit: updated output from stacktrace to (file:line function [addr])

MartinNowak commented:
Great stuff.
https://trello.com/c/FbuWfpVE/54-backtraces-with-line-numbers

MartinNowak commented:
> @nogc

That would be important, stack traces must not use the GC.

MartinNowak commented:
> add tests

As simple as adding this [here](https://github.com/D-Programming-Language/druntime/tree/master/test/exceptions) and diffing the output like [here](https://github.com/D-Programming-Language/druntime/blob/f7b582b7d1506fa4b3dad513242f34a8c940ff81/test/coverage/Makefile#L17).

MartinNowak commented:
> [0x432c37] pure @safe bool std.exception.enforceEx!(std.format.FormatException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) @ /home/administrator/Projects/d-devel/phobos/std/exception.d:589

I think the order should be `file:line function [addr]` instead.

```
/home/administrator/Projects/d-devel/phobos/std/exception.d:589 pure @safe bool std.exception.enforceEx!(std.format.FormatException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) [0x432c37]
```

It would be great to abbreviate the demangled names somehow (@ibuclaw any idea?), but it's not part of this PR.

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+
+        static if (includeBacktraceSymbols)
+        {
+            char[1024] symbolBuffer;
```
`char[1024] symbolBuffer = void`

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+        // there is an optimization that allows resolving
+        // all at once if they are sorted
+
+        char[1536] buffer; buffer[0] = 0;
```
`char[1536] buffer = void;`

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+/**
+ * This code handles backtrace generation using dwarf .debug_line section
+ * in ELF files for linux.
```
Anything linux specific about this?

yazd replied:
In the dwarf module, the pattern of the output of `backtrace_symbol` function. In the elf module, the linux elf headers and usage of `/proc/self/exe`.
But most importantly, I didn't test on other systems.

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+            machine.line += linc;
+
+            // trace("special ", ainc, " ", linc);
+            if (!callback(machine.address, LocationInfo(machine.fileIndex, machine.line))) return true;
```
Looks like it could be easily turned into a Range instead of using the callback scheme, but that's secondary.

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+{
+    assert(addresses.length > 0);
+
+    size_t addressId = 0;
```
Seems like this is never changed.

yazd replied:
It is changed on lines 206 and 208.

MartinNowak commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,464 @@
+private:
+
+// addresses must be sorted
+void resolveAddresses(ElfSection* debugLineSection, const(size_t)[] addresses, scope void delegate(Location location) callback)
```
Maybe just turn into `resolveAddress` then.

yazd replied:
It can be simplified into `resolveAddress` for the time being as this is how it is used currently. I think I'll do this.

ZombineDev commented:
Great! This is a step in the right direction.

> I think the order should be file:line function [addr] instead.
> ...
> It would be great to abbreviate the demangled names somehow...

@MartinNowak, to abbreviate the demangled names we can use:
+ eponymous templates if possible (when template and function names match) - there's no ambiguity, so DRY.

And omit:
+ Return type:  file + line is enough to remove ambiguity.
+ Function attributes: there's so ambiguity and attributes don't provide any additional info at this point.
+ Fully qualified function name (full name only for template args): file name is enough, so DRY.
+ Runtime function params (only tempate params): again, file + line is enough to remove ambiguity.
+ Anything before _Dmain, if the exception is not actually in druntime.

With my and your proposed changes it should look something like this:
```d
std.format.FormatException@/home/administrator/Projects/d-devel/phobos/std/format.d(487): Orphan format specifier: %s
----------------
/home/administrator/Projects/d-devel/phobos/std/exception.d:589 enforceEx!(std.format.FormatException, bool)(..) [0x432c37]
/home/administrator/Projects/d-devel/phobos/std/format.d:487 formattedWrite!(std.array.Appender!(immutable(char)[]), char)(..) [0x430089]
/home/administrator/Projects/d-devel/phobos/std/format.d:6548 format!(char)(..) [0x42faf3]
foo/bar.d:9 f1!(main.main().__lambda1(int))(..) [0x42fa92]
foo/bar.d:19 f2!(main.main().__lambda1(int))(..) [0x4382bc]
foo/bar.d:13 f1!(main.main().__lambda1(int))(..) [0x42faa4]
foo/bar.d:19 f2!(main.main().__lambda1(int))(..) [0x4382bc]
foo/bar.d:13 f1!(main.main().__lambda1(int))(..) [0x42faa4]
main.d:10 _Dmain [0x42fa65]
```

Additionally, we can pretty print the backtrace by:
+ Reversing the order: _Dmain first (at the top) -> throw location last (at the bottom) - so there's no need to scroll.
+ Grouping the functions by package.
+ Omitting the unnecessary .d file extention.
+ Placing the template instantiation arguments after the function name and address.
+ Aligning the function names and addresses in each package.

For example:

```d
in package main (./):
	at main:10 _Dmain(..) [0x42fa65]

in package foo (./foo):
	at bar:13 f1!(..)(..) [0x42faa4] instanciated with: main().__lambda1(int)
	at bar:19 f2!(..)(..) [0x4382bc] instanciated with: main().__lambda1(int)
	at bar:13 f1!(..)(..) [0x42faa4] instanciated with: main().__lambda1(int)
	at bar:19 f2!(..)(..) [0x4382bc] instanciated with: main().__lambda1(int)
	at bar:9  f1!(..)(..) [0x42fa92] instanciated with: main().__lambda1(int)

in package std (/home/administrator/Projects/d-devel/phobos/std):
	at format:6548:    format!(..)(..)		   [0x42faf3] instanciated with: (char)
	at format:487:     formattedWrite!(..)(..) [0x430089] instanciated with: (std.array.Appender!(immutable(char)[]), char)
	at exception:6548: enforceEx!(..)(..)	   [0x432c37] instanciated with: (std.format.FormatException, bool)

-------------------
std.format.FormatException@/usr/include/dmd/phobos/std/format.d(487): Orphan format specifier: %s
```

I want to work on this (and this is totally outside the scope of this PR), but I won't be able to probably until mid-September. Should I add myself to the trello card?

yazd commented:
updated

- [x] changed order to file:line function [addr]
- [x] added @nogc and nothrow
- [x] fixed indentation
- [x] void initialized buffers
- [ ] add tests
- [ ] fix non-linux and 32-bit builds
- [ ] simplify `resolveAddresses` into `resolveAddress`

yazd commented:
I wanted to abbreviate function names but I don't know of a way to do it easily.
@ZombineDev, I agree with all of your suggestions except for the grouping by package. But let's keep this discussion for the next PR.

MartinNowak commented:
> Omitting the unnecessary .d file extention.

That one is a false friend. What you want is to hightlight the filename and line number to quickly open that in your editor, e.g. `/home/administrator/Projects/d-devel/phobos/std/format.d:6548`.

MartinNowak commented:
Grouping by package sounds nice but might not work well in callback heavy code.

ibuclaw commented:
> It would be great to abbreviate the demangled names somehow

Omit the return type and attributes from the demangle would be a start.  Can core.demangle even do that?  Not sure about abbreviating templates, but it sounds like could turn into an art form.

MartinNowak commented:
> Not sure about abbreviating templates, but it sounds like could turn into an art form.

So be it, but the ideas above already look great, and it's also a useful functionality to have for core.demangle.

MartinNowak commented:
@yazd, do you need help with the tests or anything else?

yazd commented:
I am good. But I do wonder if I should version out all of the code for the other OS'es.
I will have time to do the tests and all that's remaining tomorrow if not today.

MartinNowak commented:
> src/rt/backtrace/elf.d(172): Error: undefined identifier 'Elf32_Ehdr', did you mean alias 'Elf_Ehdr'?

Might want to put a `version(linux):` at the top of the file. I'll try to follow up with FreeBSD and OSX support.

yazd commented:
Adding `FreeBSD` support should be easy as it uses ELF. To support `OSX`, a `Mach-O` reader is needed.

MartinNowak commented:
> To support OSX, a Mach-O reader is needed.

Ah right, forgot about that detail ;-).

yazd commented:
I have added a test and I needed to use sed to make the diff not depend on the addresses (is there a problem with depending on sed?). The test works on 64-bit, but not on 32-bit as there seems to be a bug in resolving addresses under 32-bit. I will download a 32-bit distro and test under a VM tomorrow night.

yazd commented:
Btw, I added @nogc and nothrow to everything but to the `opApply` implementation due to two reasons:
* `core.demangle.demangle` is not @nogc, nothrow.
* `opApply` takes a delegate that is not necessarily `@nogc` and calls it.

The current state is similar to the old one.

MartinNowak commented:
> opApply takes a delegate that is not necessarily @nogc and calls it.

Just as a side note, you can overload opApply for different delegates or templatize the function and let the compiler infer @nogc and nothrow.

> I will download a 32-bit distro and test under a VM tomorrow night.

You can simply do `make MODEL=32 -f posix.mak unittest`.

yazd commented:
All green. Please let me squash the commits once the review is done.

MartinNowak commented:
> As I have noted in the main post: --export-dynamic is used to get the function name. Currently .debug_line is the only section parsed and it only provides file and line number information. There is another section .debug_info that can provide the function name without --export-dynamic. I guess that will be my next task if it is needed. :P

We already rely on `--export-dynamic` for our current backtraces so it won't be an issue, but getting away w/o it would be preferred.

MartinNowak commented on src/core/runtime.d:
```diff
@@ -509,39 +511,48 @@ Throwable.TraceInfo defaultTraceHandler( void* ptr = null )
-                    if( ret )
-                        break;
+                    import rt.backtrace.dwarf;
+                    return traceHandlerOpApplyImpl!(FIRSTFRAME, true)(callstack[0 .. numframes], dg);
```
No need to make FIRSTFRAME a template parameter, simply `traceHandlerOpApplyImpl!(true)(callstack[FIRSTFRAME .. min(FIRSTFRAME, numframes)], dg);` should work as well.

yazd replied:
:+1:

yazd commented:
I have been thinking. I had a goal of minimizing memory allocation during the trace generation process and I think I went too far. Currently the state machine (that produces the address to line mappings) is run for each address. I think it would be better to store the mappings and query against them even if it would require dynamic memory allocation. What do you think?

yazd commented:
I cleaned up the code. And reduced the number of times the state machine runs by doing all address checks in the callback.
Anything left to address?

ibuclaw commented:
Importing an rt module into a core module?

yazd commented:
Where should this go then? :D

ibuclaw commented:
Depends on why you need the import.

In any case rt is supposed to differ between compiler vendors, so it is not a good idea to add dependencies to it from common modules.

yazd commented:
The only thing I am using from `rt` is `rt.util.container.array`. The other stuff is the new code.

JakobOvrum commented on src/rt/backtrace/dwarf.d:
```diff
@@ -0,0 +1,492 @@
+            snprintf(buffer.ptr, buffer.length, "%s[0x%x]", addressBuffer.ptr, callstack[i]);
+        }
+
+        auto output = buffer[0 .. strlen(buffer.ptr)];
```
This could probably use the return value of `snprintf` instead of searching for the null terminator?

yazd replied:
:+1:

MartinNowak commented on src/core/runtime.d:
```diff
@@ -509,39 +511,59 @@ Throwable.TraceInfo defaultTraceHandler( void* ptr = null )
-                    ret = dg( pos, buf );
-                    if( ret )
-                        break;
+                    import rt.backtrace.dwarf;
```
@ibuclaw is right, you can't import an `rt` module from a `core` module. You can use an `externDFunc` binding, like [here](https://github.com/D-Programming-Language/druntime/blob/de5a0bc74c2027afc56d2326130655ad01ad5270/src/object.d#L217).

MartinNowak commented on test/exceptions/Makefile:
```diff
@@ -1,6 +1,11 @@
 include ../common.mak
 
 TESTS:=stderr_msg unittest_assert
+ifeq ($(OS)-$(BUILD),linux-debug)
```
Will the ELF backtrace fail w/o debug info? In that case it should fallback to the old .dynamic based symbols.

MartinNowak replied:
Ah I see, you already combine the location info with backtrace_symbols, so it should be fine.

yazd replied:
Without `-g`, the output will look like:
`??:? symbol name [address]`

yazd commented:
Updates:
- use output from `snprintf` instead of `strlen`.
- use `externDFunc`.

yazd commented:
And always enabled using `backtrace_symbols` (instead of having it a compile-time option) so as to use `externDFunc`.

yazd commented:
Please do not merge. I have found a bug in resolving the addresses.

MartinNowak commented:
> Please do not merge. I have found a bug in resolving the addresses.

OK, the rest LGTM.

yazd commented:
Fixed bug + all green.

yazd commented:
Anything left to address?

MartinNowak commented:
Sorry for the delay, been busy with the betas.

yazd commented:
No problem. Take your time.
The tester failure looks unrelated to this PR and it has happened to different pull requests.

MartinNowak commented:
It's a lot nicer than what we have currently but it's still far from a readable stracktrace.
```
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/data/json.d:987 const(pure @safe void function(immutable(char)[])) vibe.data.json.Json.checkType!(immutable(char)[]).checkType [0x7190b5]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/data/json.d:451 vibe.data.json.Json.get!(immutable(char)[]).getinout(@property inout(immutable(char)[]) function()) [0x718f9c]
source/app.d:46 app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]) [0x70c3ba]
/home/dawg/Code/D/DPL/dmd/src/../../phobos/std/algorithm/iteration.d:549 @property app.Track std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult.front() [0x71f753]
/home/dawg/Code/D/DPL/dmd/src/../../phobos/std/algorithm/iteration.d:988 ref std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult.__ctor(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult) [0x71fb81]
/home/dawg/Code/D/DPL/dmd/src/../../phobos/std/algorithm/iteration.d:947 std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult std.algorithm.iteration.__T6filterS273app7resolveFAyaZ9__lambda5Z.filter!(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult).filter(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult) [0x71fb2a]
source/app.d:81 std.typecons.Tuple!(immutable(char)[], app.Track[]).Tuple app.resolve(immutable(char)[]) [0x70cad0]
source/app.d:146 void app.download(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse) [0x70d019]
/home/dawg/Code/D/DPL/dmd/src/../../phobos/std/functional.d-mixin-1198:1209 void std.functional.DelegateFaker!(void function(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse)*).DelegateFaker.doIt(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse) [0x7e8c9f]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/router.d:270 _D4vibe4http6router9URLRouter13handleRequestMFC4vibe4http6server17HTTPServerRequestC4vibe4http6server18HTTPServerResponseZ21__T9__lambda3TmTAAyaZ9__lambda3MFmMAAyaZv [0x79153a]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/router.d:653 const(void function(immutable(char)[], scope void delegate(ulong, scope immutable(char)[][]))) vibe.http.router.MatchTree!(vibe.http.router.Route).MatchTree.doMatch [0x792879]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/router.d:586 void vibe.http.router.MatchTree!(vibe.http.router.Route).MatchTree.match(immutable(char)[], scope void delegate(ulong, scope immutable(char)[][])) [0x792108]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/router.d:263 void vibe.http.router.URLRouter.handleRequest(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse) [0x7912de]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/server.d:1711 bool vibe.http.server.handleRequest(vibe.core.stream.Stream, vibe.core.net.TCPConnection, vibe.http.server.HTTPListenInfo, ref vibe.http.server.HTTPServerSettings, ref bool) [0x81ab24]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/server.d:1482 void vibe.http.server.handleHTTPConnection(vibe.core.net.TCPConnection, vibe.http.server.HTTPListenInfo) [0x8192de]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/http/server.d:1376 void vibe.http.server.listenHTTPPlain(vibe.http.server.HTTPServerSettings).doListen(vibe.http.server.HTTPListenInfo, bool).__lambda3(vibe.core.net.TCPConnection) [0x818d01]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/core/drivers/libevent2_tcp.d:546 void vibe.core.drivers.libevent2_tcp.ClientTask.execute() [0x7f353a]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/core/core.d:459 void vibe.core.core.makeTaskFuncInfo!(void delegate()).makeTaskFuncInfo(ref void delegate()).callDelegate(vibe.core.core.TaskFuncInfo*) [0x73fa29]
../../../../.dub/packages/vibe-d-0.7.25-rc.1/source/vibe/core/core.d:969 void vibe.core.core.CoreTask.run() [0x73daa8]
??:? void core.thread.Fiber.run() [0x8f6eb9]
??:? fiber_entryPoint [0x8f6d9b]
??:? [0xffffffff]
```
```
./app(vibe.data.json.Json.get!(immutable(char)[]).getinout(@property inout(immutable(char)[]) function())+0x59) [0x6ed6b1]
./app(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]])+0x147) [0x6e38ff]
./app(@property app.Track std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult.front()+0x68) [0x6f41a0]
./app(ref std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult.__ctor(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult)+0x42) [0x6f45da]
./app(std.algorithm.iteration.__T12FilterResultS273app7resolveFAyaZ9__lambda5TS3std9algorithm9iteration100__T9MapResultS56_D3app10resolveURLFHAyaS4vibe4data4json4JsonZS3app5TrackTAHAyaS4vibe4data4json4JsonZ9MapResultZ.FilterResult std.algorithm.iteration.__T6filterS273app7resolveFAyaZ9__lambda5Z.filter!(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult).filter(std.algorithm.iteration.MapResult!(app.Track app.resolveURL(vibe.data.json.Json[immutable(char)[]]), vibe.data.json.Json[immutable(char)[]][]).MapResult)+0x53) [0x6f4583]
./app(std.typecons.Tuple!(immutable(char)[], app.Track[]).Tuple app.resolve(immutable(char)[])+0x55b) [0x6e3f8b]
./app(void app.download(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse)+0x7e) [0x6e44d6]
./app(void std.functional.DelegateFaker!(void function(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse)*).DelegateFaker.doIt(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse)+0x64) [0x7c2a40]
./app(_D4vibe4http6router9URLRouter13handleRequestMFC4vibe4http6server17HTTPServerRequestC4vibe4http6server18HTTPServerResponseZ21__T9__lambda3TmTAAyaZ9__lambda3MFmMAAyaZv+0x1fb) [0x76830f]
./app(const(void function(immutable(char)[], scope void delegate(ulong, scope immutable(char)[][]))) vibe.http.router.MatchTree!(vibe.http.router.Route).MatchTree.doMatch+0x196) [0x769686]
./app(void vibe.http.router.MatchTree!(vibe.http.router.Route).MatchTree.match(immutable(char)[], scope void delegate(ulong, scope immutable(char)[][]))+0x93) [0x768eeb]
./app(void vibe.http.router.URLRouter.handleRequest(vibe.http.server.HTTPServerRequest, vibe.http.server.HTTPServerResponse)+0x196) [0x7680aa]
./app(bool vibe.http.server.handleRequest(vibe.core.stream.Stream, vibe.core.net.TCPConnection, vibe.http.server.HTTPListenInfo, ref vibe.http.server.HTTPServerSettings, ref bool)+0x17d9) [0x7f5211]
./app(void vibe.http.server.handleHTTPConnection(vibe.core.net.TCPConnection, vibe.http.server.HTTPListenInfo)+0x1a3) [0x7f3933]
./app(void vibe.http.server.listenHTTPPlain(vibe.http.server.HTTPServerSettings).doListen(vibe.http.server.HTTPListenInfo, bool).__lambda3(vibe.core.net.TCPConnection)+0x36) [0x7f3332]
./app(void vibe.core.drivers.libevent2_tcp.ClientTask.execute()+0x487) [0x7cd85b]
./app(void vibe.core.core.makeTaskFuncInfo!(void delegate()).makeTaskFuncInfo(ref void delegate()).callDelegate(vibe.core.core.TaskFuncInfo*)+0x72) [0x7153a6]
./app(void vibe.core.core.CoreTask.run()+0x129) [0x713359]
./app(void core.thread.Fiber.run()+0x2a) [0x8cc9ca]
./app(fiber_entryPoint+0x61) [0x8cc8b9]
```

MartinNowak commented:
Works nicely.
The code style is somewhat different than druntime, but that's no problem.
Thanks for the great work Yazan.

MartinNowak commented:
Oh, of course we should mention that in the changelog, see #1387.

yazd commented:
Thanks for the review. I will work on adding FreeBSD support in the next few days.

MartinNowak commented:
> Thanks for the review. I will work on adding FreeBSD support in the next few days.

Nice, let me know if you need help.

yazd commented:
I've just pushed a FreeBSD follow up PR: https://github.com/D-Programming-Language/druntime/pull/1399.
